crates/valence_command/src/arg_parser.rs

```diff
@@ -0,0 +1,970 @@
+        Ok(string)
+    }
+
+    fn len() -> ArgLen {
+        ArgLen::Infinite
+    }
+
+    fn display() -> Parser {
+        Parser::String(StringArg::GreedyPhrase)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct QuotableString;
+
+impl CommandArg for QuotableString {
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn len() -> ArgLen {
+        ArgLen::Within('"')
+    }
+
+    fn display() -> Parser {
+        Parser::String(StringArg::QuotablePhrase)
+    }
+}
+
+// TODO: impl Enity properly
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct EntitySelector;
+
+impl CommandArg for EntitySelector {
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::Entity {
+            only_players: false,
+            single: false,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct SingleEntitySelector;
+
+impl CommandArg for SingleEntitySelector {
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::Entity {
+            only_players: false,
+            single: true,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub enum PlayerSelector {
+    Single(String),
+    #[default]
+    All,
+    SelfPlayer,
+    Nearest,
+}
+
+impl CommandArg for PlayerSelector {
+    type Result = PlayerSelector;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(if string == "@a" {
+            PlayerSelector::All
+        } else if string == "@s" {
+            PlayerSelector::SelfPlayer
+        } else if string == "@p" {
+            PlayerSelector::Nearest
+        } else {
+            PlayerSelector::Single(string)
+        })
+    }
+
+    fn display() -> Parser {
+        Parser::Entity {
+            only_players: true,
+            single: false,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct SinglePlayerSelector;
+
+impl CommandArg for SinglePlayerSelector {
+    type Result = PlayerSelector;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(if string == "@s" {
+            PlayerSelector::SelfPlayer
+        } else if string == "@p" {
+            PlayerSelector::Nearest
+        } else {
+            PlayerSelector::Single(string)
+        })
+    }
+
+    fn display() -> Parser {
+        Parser::Entity {
+            only_players: true,
+            single: true,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PossiblyRelative<T> {
```

Comment:
nit: Would prefer the name `AbsoluteOrRelative` (first pick) or `MaybeRelative`.